provers/stark/src/prover.rs

```diff
@@ -104,727 +98,796 @@ where
     }
 }
 
-fn apply_permutation<F: IsFFTField>(vector: &mut Vec<FieldElement<F>>, permutation: &[usize]) {
-    assert_eq!(
-        vector.len(),
-        permutation.len(),
-        "Vector and permutation must have the same length"
-    );
+pub trait IsStarkProver {
+    type Field: IsFFTField;
 
-    let mut temp = Vec::with_capacity(vector.len());
-    for &index in permutation {
-        temp.push(vector[index].clone());
+    fn batch_commit(
+        vectors: &[Vec<FieldElement<Self::Field>>],
+    ) -> (BatchedMerkleTree<Self::Field>, Commitment)
+    where
+        FieldElement<Self::Field>: Serializable,
+    {
+        let tree = BatchedMerkleTree::<Self::Field>::build(vectors);
+        let commitment = tree.root;
+        (tree, commitment)
```

Comment:
maybe I'm missing some detail, but couldn't we just return the tree directly and use `tree.root` outside the function when needed?

Reply:
Ah, we can do that. But we lose the semantics of that function, because we wouldn't be commiting anymore, just building a merkle tree. We would have to rename it.

Reply:
ok maybe not too important for now, let's leave it like this

Reply:
Name is tricky, it's doing a commitment and returning the tree for further use